python/lsst/ip/isr/crosstalk.py

```diff
@@ -659,139 +686,171 @@ def subtractCrosstalk(self, thisExposure, sourceExposure=None, crosstalkCoeffs=N
+        subtrahend = targetMaskedImage.Factory(targetMaskedImage.getBBox())
         subtrahend.set((0, 0, 0))
 
-        coeffs = coeffs.transpose()
-        valid = valid.transpose()
-        # Apply NL coefficients
-        if doSqrCrosstalk:
-            coeffsSqr = coeffsSqr.transpose()
-            mi2 = mi.clone()
-            mi2.scaledMultiplies(1.0, mi)
-
-        for ss, sAmp in enumerate(sourceDetector):
-            if fullAmplifier:
-                sImage = subtrahend[sAmp.getBBox() if isTrimmed else sAmp.getRawBBox()]
-            elif parallelOverscan:
-                if detectorConfig is not None:
-                    ampConfig = detectorConfig.getOverscanAmpConfig(sAmp)
-                    if not ampConfig.doParallelOverscanCrosstalk:
-                        # Skip crosstalk correction for this amplifier.
-                        continue
+        # If we are ignoring variance and doing subtrahend masking, then
+        # we can work on only the image plane (3x speed increase).
+        imageOnly = ignoreVariance and doSubtrahendMasking
 
-                sImage = subtrahend[sAmp.getRawParallelOverscanBBox()]
-            else:
-                sImage = subtrahend[sAmp.getBBox() if isTrimmed else sAmp.getRawDataBBox()]
-            for tt, tAmp in enumerate(detector):
-                # Skip 0.0 and invalid coefficients.
-                if coeffs[ss, tt] == 0.0 or not valid[ss, tt]:
+        for sourceIndex, sourceAmp in enumerate(sourceDetector):
+            for targetIndex, targetAmp in enumerate(targetDetector):
+                coeff = coeffs[sourceIndex, targetIndex]
+                coeffSqr = coeffsSqr[sourceIndex, targetIndex]
+
+                if coeff == 0.0:
                     continue
-                if badAmpDict is not None:
-                    if badAmpDict[sAmp.getName()] or badAmpDict[tAmp.getName()]:
-                        continue
-                tImage = self.extractAmp(
-                    mi,
-                    tAmp,
-                    sAmp,
-                    isTrimmed=isTrimmed,
-                    fullAmplifier=fullAmplifier,
-                    parallelOverscan=parallelOverscan,
-                )
-                tImage.getMask().getArray()[:] &= crosstalk  # Remove all other masks
-                tImage -= backgrounds[tt]
-                sImage.scaledPlus(coeffs[ss, tt], tImage)
-                # Add the nonlinear term
-                if doSqrCrosstalk:
-                    # Note that mi2 is the square of the masked image.
-                    tImageSqr = self.extractAmp(
-                        mi2,
-                        tAmp,
-                        sAmp,
+
+                targetBBox = self._getAppropriateBBox(targetAmp, isTrimmed, fullAmplifier)
+
+                # The extractAmp() method will always make a copy of the source
+                # amplifier data.
+                if imageOnly:
+                    sourceImage = self.extractAmp(
```

Comment:
Does `extractAmp` need to be inside this if/else?  This also seems like it can be merged with the subsequent `if not imageOnly` block.

Reply:
Partly this is because of the way I wrote the code originally, but the option is to either make a temporary variable that is sent to `extractAmp` or do it this way.  I think it's a wash.

As for whether the following `if not imageOnly` is needed ... no it is not.